saiph/serializer.py

```diff
@@ -38,12 +48,51 @@ def default(self, obj):
 
         if isinstance(obj, pd.Series):
             data = obj.to_json(orient="index", default_handler=str)
-            return dict(__series__=data, dtype=str(obj.dtype))
+
+            # We cast to string and not to object, to transform the
+            # 'inferred_type' property (read-only) to string
+            # If we don't it remains at integer when using an index with `'0'`
+            index_type = (
+                "str" if str(obj.index.dtype) == "object" else str(obj.index.dtype)
+            )
+
+            return dict(__series__=data, dtype=str(obj.dtype), index_type=index_type)
 
         if isinstance(obj, pd.DataFrame):
             # orient='table' includes dtypes but doesn't work
+            dtypes = valmap(str, obj.dtypes.to_dict())
+
+            # value_types is the type of a single cell in a column,
+            # which can be different from dtypes, especially when the dtype is 'object'
+            # or when we have nan-values together with integer values or stringified integers
+            value_types = {}
+            for col in obj.columns:
+                non_null_series = obj[col].dropna()
+                value_types[col] = (
+                    type(non_null_series.values[0]).__name__
+                    if not non_null_series.empty
+                    else "float"
+                )
             data = obj.to_json(orient="index", default_handler=str)
-            return dict(__frame__=data)
+
+            # We cast to string and not to object, to transform the
+            # 'inferred_type' property (read-only) to string
+            # If we don't it remains at integer when decoding and using
+            # an index or column with `'0'`
+            index_type = (
+                "str" if str(obj.index.dtype) == "object" else str(obj.index.dtype)
+            )
```

Comment:
Aren't you doing this twice ?

Reply:
Yes, but I need to be inside the if-statements so that other types of objects (e.g. str) do not execute that line

Reply:
Yes, but I need to be inside the if-statements so that other types of objects (e.g. str) do not execute that line